k3s-io/k3s
# Fix to etcd-snapshot ls and other commands always creating the etcd snapshot folder

#### Problem ####
Currently when running k3s as root and no snapshots, calling the command `etcd-snapshot ls` with a non-root user **will not** return an error that `/var/lib/rancher/k3s/server/db/snapshots` permission denied. Instead, the command will create a completely useless and empty `~/.rancher/k3s/server/db/snapshots` folder structure because the command was not run as root.

#### Proposed Changes ####
Only create the snapshot folder structure on 3 commands:
1. Snapshot save
2. S3 Download
3. StoreSnapshotData

For the other commands: list, delete, S3 upload, prune... do not create a folder. If the folder does not exist, err out. 
So if no snapshots exist and you call `sudo k3s etcd-snapshot delete TESTER` the command will return a directory does not exist error.
If you call `k3s etcd-snapshot ls` as a non-root when snapshots do exist under `/var/lib/rancher/k3s...` it will return a directory does not exist error. 

#### Types of Changes ####
Bugfix


#### Verification ####
1. Start up k3s cluster with etcd as root
2. `sudo k3s etcd-snapshot save`
2. As a non-root user, call `k3s etcd-snapshot ls`... you should get a directory does not exist error
3. Check that `~/.rancher/k3s/` folder structure has not been created. 
4. `sudo k3s etcd-snapshot ls` should list the snapshots correctly

#### Linked Issues ####
https://github.com/k3s-io/k3s/issues/3745


#### User-Facing Change ####

```release-note
Users must create or download etcd-snapshots before being able to list or delete snapshots. 
```

#### Further Comments ####
